source/common/router/scoped_rds.cc

```diff
@@ -453,9 +453,10 @@ ScopedRdsConfigSubscription::detectUpdateConflictAndCleanupRemoved(
       exception_msg = fmt::format("duplicate scoped route configuration '{}' found", scope_name);
       return clean_removed_resources;
     }
-    const envoy::config::route::v3::ScopedRouteConfiguration& scoped_route_config =
+    envoy::config::route::v3::ScopedRouteConfiguration scoped_route_config =
         scope_config_inserted.first->second;
-    const uint64_t key_fingerprint = MessageUtil::hash(scoped_route_config.key());
+    const uint64_t key_fingerprint =
+        ScopedRouteInfo(std::move(scoped_route_config), nullptr).scopeKey().hash();
```

Comment:
Could we use a strong int type for the hash to avoid incorrect comparison of hashes computed by different mechanisms?  Is there potential for similar bugs to exist in other places where Envoy compares config fingerprints?

Reply:
ScopedKey is special, it implements its own hashing for several reasons.

Reply:
Right.  How can we avoid accidentally comparing the hash of apples vs the hash of oranges?

Reply:
A strong hash type may be beyond the scope of this pull request, but we can discuss it later.

Reply:
scopeKey::hash() is unique if the fragments definitions are different. 

using a strong type sounds good to me.  @chaoqin-li1123 I think we can just define a new type which wraps around an int64_t

Reply:
@chaoqin-li1123 I think it would be a good change to make ScopeKey::hash() return a struct that wraps `uint64_t`. This will blow that change somewhat because you will need to fix the cases where it is used in hash tables. I think it is easy to fix by following this example: https://github.com/abseil/abseil-cpp/blob/037ade20d1132781aae3cda4d547a9e6a5f557bf/absl/hash/hash.h#L49

Reply:
I'm not seeing a strong int library in either Envoy or Abseil, so use of a strong int here seems more challenging than I would have expected.  I guess this is something to consider in the future.  I guess some ASSERTs is as good as we're going to get for now.

Reply:
I think raw int is fine for now since it is not envoy style to use strong type int.